OpenAI: Function Calling (#7704)
mendix/docs: content/en/docs/appstore/modules/openai/function-calling.md

---
title: "Function Calling"
url: /appstore/modules/openai-connector/function-calling/
category: "Open AI"
linktitle: "Function Calling"
weight: 5
description: "Describes function calling with OpenAI"
tags: ["OpenAI", "generative AI", "AI", "connector", "marketplace", "chatgpt", "genAI", "Azure OpenAI", "function calling", "tools", "ReAct", "chat completions", "text generation"]
---

## 1 Introduction {#introduction}

Function calling enables LLMs (Large Language Models) to connect with external tools to gather information, execute actions, convert natural language in structure data and much more. Thereby, function calling enables the model to intelligently decide when to call one or more predefined functions to gather additional information to include in the assistant's response. OpenAI does not call the function; instead, the model returns a tool call JSON structure that is used to build the input of the function(s) so that they can be executed as part of the chat completions operation.

### 1.1 High-level flow {#high-level}

The basic process steps of function calling contain at least the following:

> 1. Invoke the chat completions API with a user prompt and a list of available functions with expected input parameters.

Reviewer: functions --> functions (microflows)

> 2. Based on the user prompt and the available functions, the model can intelligently decide whether to call one or more functions. If so, the content of the assistant's response will be a stringified JSON object containing the input parameters of the function as desceribed in the request. Note that the model may hallucinate parameters, so they should be validated before the function is actually called.

Reviewer: "whether to call one or more functions" --> "that it's needed to call one or more functions"

> 3. Parse the string into JSON and call the function with its input parameters.

Reviewer: function --> function (microflow)

Reviewer 2: https://github.com/mendix/docs/pull/7704/files#r1593792140

4. Append the existing list of messages with the function response as a new tool message. Then, invoke the chat completions API again and let the model answer your initial prompt with the new information provided by the function.

For more general information on this topic, see [Function Calling](https://platform.openai.com/docs/guides/function-calling).

> ### 1.2 Function calling with the OpenAI Connector {#openai-connecor}

Reviewer: Can we include that the microflows run in the original user session, and have access to `$currentUser` for applying user specific data / security constraints?

Reviewer 2: https://github.com/mendix/docs/pull/7704/files#r1593797902

> Function calling is supported for all chat completions operations by adding the optional input parameter [FunctionCollection](#functioncollection). Functions in Mendix are essentially microflows that can be registered within the request to the LLM​. The OpenAI connector takes care of handling the tool call response as well as executing the function microflow(s) until the API returns the final assistant's response. Currently, function microflows are limited to one input parameter of type string and must return a string.

Reviewer 2: This seems to contradict what we say before. If the connector calls the function with the parameters how can we check that the model hasn't hallucinated parameters before the function is called?

Reviewer 3: It can be validated inside the function microflow which needs to be implemented by the developer

{{% alert color="warning" %}}
Function calling is a very powerful capability, but this also introduces potential risks. Function microflows do not respect entity access of the current user. Make sure to only retrieve and return information that the user is allowed to view, otherwise confidential information may be visible to the current user in the assistant's response. Furthermore, we strongly advise developer to build user confirmation logic into function microflows that have a potential impact on the world on behalf of the user, such as sending an email, posting online, making a purchase, etc.
{{% /alert %}}

### 1.3 Use cases {#use-cases}

Function calling can be used for a variety of usecases including

* Create assisants that can answer questions about data from your Mendix database or a knowledge base
    * e.g., getTicketById(String identifier) or findSimilarTickets(String description)
* Create assistants that can get information from external APIs
    * e.g., getCurrentWeather(String location)
* Extract structured data from natural language

>     * e.g., extractBirthday(String birthday), or getSqlQuery(String query)

Reviewer: Let's not including SQL query as an example since that may cause OQL like expectations. Let's not do that for now.

* Execute action like creating objects

>     * e.g., createTicket(String subject)

Reviewer: Let's include a recommendation that if the function manipulates things (e.g. creating a ticket), user validation to confirm the action is recommended to prevent unexpected behavior

OpenAI does not call the function. The model returns a tool call JSON structure that is used to build the input of the function(s) so that they can be executed as part of the chat completions operation.

### 1.1 High-level flow {#high-level}

The basic steps of function calling are as follows:

1. Invoke the chat completions API with a user prompt and a list of available functions (microflows) with expected input parameters.

    * The model decides whether to call one or more functions based on the user prompt and the available functions.

        * If it calls a function, the content of the assistant's response will be a stringified JSON object containing the input parameters of the function as described in the request. Note that the model may hallucinate parameters, so they should be validated before the function is actually called.

>         * Parse the string into JSON and call the function with its input parameters. The microflow runs in the original user's session and `$currentUser` can for example be used to apply security constraints.

Reviewer 2: ```suggestion
        * Parse the string into JSON and call the function (microflow) with its input parameters. The microflow runs in the original user's session and `$currentUser` can for example be used to apply security constraints.
```

        * Append the existing list of messages with the function response as a new tool message.

> 

Reviewer 2: I don't find it clear which bits are done by the OpenAI connector and what the customer needs to do. I've tried to clarify it, based on my understanding of the rest of the document.

Here, and in the diagrams below, it seems that the OpenAI connector does everything and the customer just has to make a single call.

However, you also say that "the model may hallucinate parameters, so they should be validated before the function is actually called". But if the connector is calling the function directly you can't validate the parameters before you call the function - you can only validate the parameters inside the function after it has been called.

Reviewer 3: you can build validation inside the fucntion microflow. Even though the function microflow is called by the OpenAI connector, the function microflow is implemented by the developer and could some validation as the first action

Reviewer 3: @MarkvanMents can we go back to the structure with the 4 steps, which align with the process mentioned in the official openai docs? https://platform.openai.com/docs/guides/function-calling

Reviewer 2: I find the four steps confusing for the Mendix documentation as the developer is only interested in step 1 and step 4 while steps 2 and 3 are hidden by the wrapper. 
So my idea is to have four steps, but make it clear that steps 2 and 3 are done automatically by the connector.

I'm happy to go for another structure, but I think it should be clear what is done by the developer and what is done by the connector.

As it was it just described what the openai API does, which the developer can get from the openai docs.

Two helper microflow are available to construct the `FunctionCollection` with a list of `Functions`:

* `FunctionCollection_CreateAndAddFunction` can be used to initialize a new `FunctionCollection` and add a new `Function` to it in order to enable [function calling](/appstore/modules/openai-connector/#chatcompletions-functioncalling).
* `FunctionCollection_AddFunction` can be used to add a new `Function` to an existing `FunctionCollection`.

Depending on the user prompt and the available functions, the model may suggest multiple tool calls to the same or different functions or there might be multiple API calls followed by new tools calls until the model returns the final assistant's response.
A way to steer the function calling process is the [ToolChoice](/appstore/modules/openai-connector/#enum-toolchoice) parameter. This optional attribute on the [FunctionCollection](/appstore/modules/openai-connector/#functioncollection) entity controls which (if any) function is called by the model.

{{% alert color="warning" %}}

> Function calling is a very powerful capability, but this also introduces potential risks. Function microflows do not respect any entity access rules for the current end-user. Make sure to retrieve and return only information that the end-user is allowed to view, otherwise confidential information may be visible to the current end-user in the assistant's response.

Reviewer 2: ```suggestion
Function calling is a very powerful capability, but this also introduces potential risks. Function microflows do not respect any entity access rules for the current end-user. You can use `$currentUser` to help you ensure you retrieve and return only information that the end-user is allowed to view, otherwise confidential information may be visible to the current end-user in the assistant's response.
```

Function calling enables LLMs (Large Language Models) to connect with external tools to gather information, execute actions, convert natural language into structured data, and much more. Function calling thus enables the model to intelligently decide when to let the Mendix app call one or more predefined functions (microflows) to gather additional information to include in the assistant's response.

OpenAI does not call the function. The model returns a tool call JSON structure that is used to build the input of the function(s) so that they can be executed as part of the chat completions operation.

### 1.1 High-level flow {#high-level}

If you use the `ChatCompletions_Execute_WithoutHistory` or `ChatCompletions_Execute_WithHistory` microflows to do your chat completions, you can use the OpenAI connector to do function (microflow) calling for you in just two steps:

1. Invoke the chat completions API with a user prompt and a list of available functions (microflows) with expected input parameters.

    The model will decide which function (microflow) to call, if any, and will respond based on the information you provide and the response from any function (microflow) it calls.

> 2. Invoke the chat completions API again and let the model answer your initial prompt with the new information provided by the function.

Reviewer 2: Does the developer even have to do this? From the examples in section 1.4 it looks like the connector does this step too and basically the function call is a single-step process.

Reviewer 3: for with and without history operations, this is handled by the connector. For advanced it isn't

Reviewer 2: OK, I'll remove this and just have it in the detailed description.

## 1 Introduction {#introduction}

Function calling enables LLMs (Large Language Models) to connect with external tools to gather information, execute actions, convert natural language into structured data, and much more. Function calling thus enables the model to intelligently decide when to let the Mendix app call one or more predefined functions (microflows) to gather additional information to include in the assistant's response.

OpenAI does not call the function. The model returns a tool call JSON structure that is used to build the input of the function(s) so that they can be executed as part of the chat completions operation.

### 1.1 High-level flow {#high-level}

The basic steps of function calling are as follows:

1. Invoke the chat completions API with a user prompt and a list of available functions (microflows) with expected input parameters.

    * The model decides whether to call one or more functions based on the user prompt and the available functions.

>         * If it calls a function, the content of the assistant's response will be a stringified JSON object containing the input parameters of the function as described in the request. Note that the model may hallucinate parameters, so they should be validated before the function is actually called.

Reviewer 2: What is the response if it doesn't call a function?
What is the format of the JSON object? I can't see an example anywhere and so I'm not sure how I know which function was called, what the parameters were, and what the function returned before the AI parses it and sends it back.

The examples below do not return the JSON to the customer, they just return a natural language response ("The status of ticket 123 is in progress") so I'm not sure why the customer needs to know about the JSON or what they can do to get it.

Reviewer 3: the format of the JSON object was provided as part of the request so it depends on the input of your function microflow.
This is just some background information for developers who want to know how function calling works under the hood. Developers might need to handle the json object and the function call themselves if they are using the advanced operations, which we generally don't recommend. If they are using with/without history operations, we handle everything inside the connector

Reviewer 2: I'm finding this confusing. Are we describing here how function calls work, or just how the OpenAI connector does it for you?

If developers are doing the calls themselves, then we should be able to point them at the OpenAI documentation. I think we should focus here on how the non-advanced calls work, or at least make it clear what the differences are. Here we are mixing the two things up which, as someone who doesn't know the internal workings of the connector, is confusing.

Additionally, we don't anywhere describe what JSON we create for a request, we only describe entities where the information is held, so the developer isn't going to know the structure.

I think that if we mention the JSON objects we should have a bit more information about them?

Reviewer 2: In any case, I think we should be clearer where the validation takes place:
```suggestion
        * If it calls a function, the content of the assistant's response will be a stringified JSON object containing the input parameters of the function as described in the request. Note that the model may hallucinate parameters, so they should be validated within the function before being used.
```
